graal-js/src/com.oracle.truffle.js.test/src/com/oracle/truffle/js/test/builtins/PromiseRejectionHandler.java

```diff
@@ -0,0 +1,120 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.oracle.truffle.js.test.builtins;
+
+import com.oracle.truffle.js.test.JSTest;
+import org.graalvm.polyglot.Context;
+import org.graalvm.polyglot.Value;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+public class PromiseRejectionHandler {
+
+    @Test
+    public void testExists() {
+        try (Context context = newUnhandledRejectionContext()) {
+            assertTrue(registerHandlerFunctionExists(context));
+        }
+        try (Context context = JSTest.newContextBuilder().build()) {
+            assertFalse(registerHandlerFunctionExists(context));
+        }
+    }
+
+    @Test
+    public void testInvokeRegister() {
+        try (Context context = newUnhandledRejectionContext()) {
+            context.eval("js", "Graal.registerPromiseRejectionHandler(() => {}); Promise.reject('test')");
+            context.eval("js", "Graal.registerPromiseRejectionHandler(); Promise.reject('test')");
+            context.eval("js", "Graal.registerPromiseRejectionHandler(123); Promise.reject('test')");
+        }
+    }
+
+    @Test
+    public void testFireSimple() {
+        try (Context context = newUnhandledRejectionContext()) {
+            context.eval("js", "count = 0; Graal.registerPromiseRejectionHandler(() => count++); Promise.reject('test')");
+            Value count = context.eval("js", "count");
+            assertEquals(1, count.asInt());
+        }
+    }
+
+    @Test
+    public void testUnhandledRejectionInHandler() {
+        try (Context context = newUnhandledRejectionContext()) {
+            context.eval("js", "" +
+                    "count = 0; " +
+                    "Graal.registerPromiseRejectionHandler(() => {" +
+                    "   if (count == 0) Promise.reject('testInner');" +
+                    "   count++;" +
+                    "}); " +
+                    "Promise.reject('test')");
+            Value count = context.eval("js", "count");
+            assertEquals(2, count.asInt());
+        }
+    }
+
+    @Test
+    public void testFireArguments() {
+        try (Context context = newUnhandledRejectionContext()) {
+            String handlerName = "Packages." + getClass().getName() + ".testFireArgumentsHandler";
+            context.eval("js", "Graal.registerPromiseRejectionHandler((r, p) => " + handlerName + "(r, p)); Promise.reject('test')");
```

Comment:
in this test here, I was directly passing the host function into the promised rejection handler. I noticed it wasn't actually being called, because I think it failed the  `isJSFunction` test here:

https://github.com/oracle/graaljs/pull/624/files#diff-056c919be57941109333065d2428bb3a12b5f10390d618512c80a7a5a5c5953aR2013-R2018

Is there a better way to test for something that is "function like" that can handle host functions?

Reply:
and if the function is a host function and passes the above test, would it fail here trying to call it as a `JSFunction`?:

https://github.com/oracle/graaljs/blob/8443e4246accd3eb373bb24348ffaa7dca8a3204/graal-js/src/com.oracle.truffle.js/src/com/oracle/truffle/js/nodes/promise/BuiltinPromiseRejectionTracker.java#L130-L133

Reply:
switched to `JSInteropUtil` for calling the function.